src/Symfony/Component/Messenger/Worker.php

```diff
@@ -282,4 +326,79 @@ public function getMetadata(): WorkerMetadata
     {
         return $this->metadata;
     }
+
+    public function handleFutures(string $transportName, $parallelProcessLimit): void
+    {
+        $toHandle = self::$futures;
+        self::$futures = [];
+
+        if (!$toHandle) {
+            return;
+        }
+
+        $futuresReceived = [];
+        $envelopesAssociated = [];
+
+        foreach ($toHandle as $combo) {
+            $futuresReceived[] = $combo[0];
+            $envelopesAssociated[] = $combo[1];
+        }
+
+        [$errorsFromAwait, $executions] = awaitAll($futuresReceived);
+
+        foreach ($errorsFromAwait as $index => $error) {
+            try {
+                $execution = $futuresReceived[$index]->await();
+                $envelope = $execution->await();
+            } catch (\Throwable $e) {
+                $this->preAck($envelopesAssociated[$index]->withoutStampsOfType(FutureStamp::class), $transportName, false, $e);
+
+                continue;
+            }
+
+            $this->preAck($envelope, $transportName, false, null);
+        }
+
+        $futures = array_map(fn ($execution) => $execution->getFuture(), $executions);
+
+        $i = 0;
+
+        while ($executions) {
```

Comment:
I feel like we can take advantage of some amp stuff to make this much more readable.

See: `Future::iterate()`, for example. I find this entire loop quite confusing. We are iterating over `$futures`, while also maintaining a copy in `$executions`. I'm not 100% sure any of this is actually doing anything in parallel.

For example, we set $i = 0, then wait for something to complete, then filter out the completed future from `$executions`, then reset `$futures`, then increment `i`, then wait for all executions to complete, then start all over again.

Like, I'm not sure what the intent is here; some comments would be helpful to explain _why_ you are doing what you are doing. To be frank, it doesn't seem to make any sense at all.

First, it appears that you want to set a processing limit, but all these fibers/processes are already running. You can't set the parallel processing limit this late in the code. Instead, you should set it in `src/Symfony/Component/Messenger/ParallelMessageBus.php:38` when you set up the worker pool. That would be where you set the limits by defining how many threads / fibers it will run concurrently. Doing this here doesn't do anything.

Secondly, the entire code can be written differently to take advantage of amp's parallelism.

```php
$futures = self::$futures; // take a copy of futures we care about
self::$futures = []; // any other fiber shouldn't watch these fibers

// loop over all tasks in the background.
async(
  function () use ($futures)) 
  {
    // iterate over futures in completion order
    foreach(Future::iterate($futures) as $idx => $future) {
      try {
        $val = $future->await();
        // do stuff with val
      } catch (\Throwable $ex) {
        // handle error of $futures[$idx] here.
      }
    }
  }
)
```

Reply:
thanks for your feedback. 
Instead of calling `awaitFirst`, I could implement the internal logic of this function in the worker code. This way, I would have access to the index of the elements, which would make it easier to unset and sort what is done or still pending. 
But it's wrap in an `Execution`class so it would be something like 
```
foreach(Future::iterate($futuresReceived) as $idx => $future) {
            try {
                $execution = $future->await();
                $execution->getFuture()->await();
                $envelope = $execution->getFuture()->await();

                unset($futuresReceived[$idx]);

                $this->preAck($envelope, $transportName, false, null);
            } catch (\Throwable $ex) {
            }
            ```

Reply:
Regarding the action, when using the parallel message bus the messages seem to be processed concurrently rather than in order. I'll add a small screenshot to show the result compared to a "normal" bus